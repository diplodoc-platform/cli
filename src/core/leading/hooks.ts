import type {LeadingPage, Plugin} from './types';

import {AsyncParallelHook, AsyncSeriesHook, AsyncSeriesWaterfallHook} from 'tapable';

import {generateHooksAccess} from '~/core/utils';

export function hooks(name: string) {
    return {
        /**
         * Emits relative to root asset path on each local link in Leading.
         * This includes paths in links and blocks sections.
         */
        Asset: new AsyncParallelHook<[RelativePath, RelativePath]>(
            ['asset', 'path'],
            `${name}.Asset`,
        ),
        Plugins: new AsyncSeriesWaterfallHook<[Plugin[]]>(['plugins'], `${name}.Plugins`),
        Resolved: new AsyncSeriesHook<[DeepFrozen<LeadingPage>, RelativePath]>(
            ['leading', 'path'],
            `${name}.Resolved`,
        ),
        Dump: new AsyncSeriesWaterfallHook<[LeadingPage, NormalizedPath]>(
            ['leading', 'path'],
            `${name}.Dump`,
        ),
    };
}

const [getHooks, withHooks] = generateHooksAccess('Leading', hooks);

export {getHooks, withHooks};
