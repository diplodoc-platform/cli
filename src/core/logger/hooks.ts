import {SyncHook} from 'tapable';

import {generateHooksAccess} from '~/core/utils';

export function hooks(name: string) {
    return {
        Info: new SyncHook<[string]>(['message'], `${name}.Info`),
        Warn: new SyncHook<[string]>(['message'], `${name}.Warn`),
        Error: new SyncHook<[string]>(['message'], `${name}.Error`),
    };
}

const [getHooks, withHooks] = generateHooksAccess('Logger', hooks);

export {getHooks, withHooks};
