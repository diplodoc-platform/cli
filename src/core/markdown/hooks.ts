import type {AdditionalInfo, Collect, Plugin} from './types';

import {AsyncSeriesHook, AsyncSeriesWaterfallHook} from 'tapable';

import {generateHooksAccess} from '~/core/utils';
import {Meta} from '~/core/meta';

export function hooks(name: string) {
    return {
        Collects: new AsyncSeriesWaterfallHook<[Collect[]]>(['collects'], `${name}.Collects`),
        Plugins: new AsyncSeriesWaterfallHook<[Plugin[]]>(['plugins'], `${name}.Plugins`),
        Loaded: new AsyncSeriesHook<[string, DeepFrozen<Meta>, NormalizedPath]>(
            ['markdown', 'meta', 'path'],
            `${name}.Loaded`,
        ),
        Resolved: new AsyncSeriesHook<[string, NormalizedPath]>(
            ['markdown', 'path'],
            `${name}.Resolved`,
        ),
        Dump: new AsyncSeriesWaterfallHook<[string, NormalizedPath, AdditionalInfo]>(
            ['markdown', 'path', 'info'],
            `${name}.Dump`,
        ),
    };
}

const [getHooks, withHooks] = generateHooksAccess('Markdown', hooks);

export {getHooks, withHooks};
