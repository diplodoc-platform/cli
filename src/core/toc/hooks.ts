import type {IncludeInfo, IncluderOptions, RawToc, RawTocItem, Toc} from './types';

import {AsyncParallelHook, AsyncSeriesWaterfallHook, HookMap} from 'tapable';

import {generateHooksAccess, intercept} from '~/core/utils';

const name = 'Toc';

export function hooks() {
    return intercept(name, {
        /**
         * Called before item data processing (but after data interpolation)
         */
        Item: new AsyncSeriesWaterfallHook<[RawTocItem, RelativePath]>(
            ['TocItem', 'TocPath'],
            `${name}.Item`,
        ),
        /**
         * AsyncSeriesWaterfall HookMap called for each includer name detected in toc.<br>
         * Expects RawToc as result of waterfall.
         */
        Includer: new HookMap(
            (type: string) =>
                new AsyncSeriesWaterfallHook<[RawToc, IncluderOptions, RelativePath]>(
                    ['Toc', 'options', 'TocPath'],
                    `${name}.Includer(${type})`,
                ),
        ),
        Resolved: new AsyncParallelHook<[Toc, RelativePath]>(
            ['Toc', 'TocPath'],
            `${name}.Resolved`,
        ),
        Included: new AsyncParallelHook<[Toc, RelativePath, IncludeInfo]>(
            ['Toc', 'TocPath', 'IncludeInfo'],
            `${name}.Included`,
        ),
    });
}

const [getHooks, withHooks] = generateHooksAccess(name, hooks);

export {getHooks, withHooks};
