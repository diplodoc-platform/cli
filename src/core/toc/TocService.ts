import type {Run as BaseRun} from '~/core/run';
import type {VarsService} from '~/core/vars';
import type {MetaService} from '~/core/meta';
import type {
    EntryTocItem,
    GraphData,
    GraphTocData,
    IncludeInfo,
    RawToc,
    RawTocItem,
    Toc,
    WithItems,
} from './types';
import type {LoaderContext} from './loader';

import {basename, dirname, join, relative} from 'node:path';
import {dump, load} from 'js-yaml';
import {dedent} from 'ts-dedent';

import {
    Defer,
    Graph,
    VFile,
    bounded,
    copyJson,
    errorMessage,
    memoize,
    normalizePath,
    own,
} from '~/core/utils';

import {getHooks, withHooks} from './hooks';
import {isMergeMode, loader} from './loader';
import {isEntryItem} from './utils';

export type TocServiceConfig = {
    ignore: string[];
    ignoreStage: string[];
    template: {
        enabled: boolean;
        features: {
            conditions: boolean;
            substitutions: boolean;
        };
        scopes: {
            code: boolean;
            text: boolean;
        };
    };
    removeHiddenTocItems: boolean;
    removeEmptyTocItems: boolean;
};

type WalkStepResult<I> = I | I[] | null | undefined;

export type WalkStepContext<T extends object = {}> = Hash<unknown> & T;

type RestrictedAccessContext = WalkStepContext<{
    'restricted-access'?: string[][];
}>;

type WalkOptions<T> = {
    accept: (item: T) => boolean;
};

type Run = BaseRun<TocServiceConfig> & {
    vars: VarsService;
    meta: MetaService;
};

type Options = {
    skipMissingVars: boolean;
    mode: 'translate' | 'build';
};

@withHooks
export class TocService {
    readonly name = 'Toc';

    readonly relations = new Graph<GraphData>();

    get tocs() {
        return (this.relations.overallOrder() as NormalizedPath[]).filter(this.isToc).map(this.for);
    }

    get entries() {
        return (this.relations.overallOrder() as NormalizedPath[]).filter(this.isEntry);
    }

    private run: Run;

    private logger: Run['logger'];

    private config: TocServiceConfig;

    private options;

    private get vars() {
        return this.run.vars;
    }

    private get meta() {
        return this.run.meta;
    }

    constructor(run: Run, options: Options = {skipMissingVars: false, mode: 'build'}) {
        this.run = run;
        this.logger = run.logger;
        this.config = run.config;
        this.options = options;
    }

    async init(paths: NormalizedPath[]) {
        for (const path of paths) {
            await this.load(path);
        }

        return this.tocs.filter((toc) => paths.includes(toc.path));
    }

    async dump(file: NormalizedPath, toc?: Toc): Promise<VFile<Toc>> {
        toc = toc || this.for(file);

        return this._dump(toc.path, toc);
    }

    /**
     * Visits items which will be project entries. Applies actor to each item.
     * Then applies actor to each item in actor result.items.
     * Returns actor results.
     */
    async walkEntries<T extends WithItems<T> & {href: NormalizedPath}>(
        items: T[] | undefined,
        actor: (item: T) => Promise<WalkStepResult<T>> | WalkStepResult<T>,
    ): Promise<T[] | undefined> {
        return this.walkItems(items, actor, {accept: isEntryItem});
    }

    /**
     * Visits all passed items. Applies actor to each item.
     * Then applies actor to each item in actor result.items.
     * Returns actor results.
     */
    async walkItems<T extends WithItems<T>>(
        items: T[] | undefined,
        actor: (
            item: T,
            context: WalkStepContext,
        ) => Promise<WalkStepResult<T>> | WalkStepResult<T>,
        options: WalkOptions<T> = {accept: () => true},
    ): Promise<T[] | undefined> {
        return this._walkItems(items, actor, options);
    }

    /**
     * Resolves toc path and data for any page path.
     * Expects what all paths are already loaded in service.
     */
    @bounded for(path: RelativePath): Toc {
        const file = normalizePath(path);

        if (this.isToc(file)) {
            return this.relations.getNodeData(file).data as Toc;
        }

        const tocPaths = (this.relations.dependantsOf(file) as NormalizedPath[]).filter(this.isToc);
        if (!tocPaths.length) {
            throw new Error('Error while finding toc dir.');
        }

        if (tocPaths.length === 1) {
            return this.relations.getNodeData(tocPaths[0]).data as Toc;
        }

        const fileParts = normalizePath(join(dirname(file), 'toc.yaml')).split('/');
        const toc = tocPaths.reduce(
            (result, path) => {
                const tocParts = path.split('/');

                let index = 0;
                let score = 0;
                while (tocParts.length > index && fileParts[index] === tocParts[index]) {
                    index++;
                    score++;
                }

                if (score > result.score) {
                    return {score, path};
                }

                return result;
            },
            {score: 0, path: null} as {score: number; path: null | NormalizedPath},
        );

        if (toc.path === null) {
            throw new Error('Error while finding toc dir.');
        }

        return this.relations.getNodeData(toc.path).data as Toc;
    }

    release(path: NormalizedPath) {
        memoize.release(this._dump, path);
    }

    @bounded isToc(path: NormalizedPath) {
        if (!this.relations.hasNode(path)) {
            return false;
        }

        const data = this.relations.getNodeData(path);

        return data.type === 'toc';
    }

    @bounded isEntry(path: NormalizedPath) {
        if (!this.relations.hasNode(path)) {
            return false;
        }

        const data = this.relations.getNodeData(path);

        return data.type === 'entry';
    }

    @bounded isGenerator(path: NormalizedPath) {
        if (!this.relations.hasNode(path)) {
            return false;
        }

        const data = this.relations.getNodeData(path);

        return data.type === 'generator';
    }

    @memoize('path')
    private async _dump(file: NormalizedPath, toc: Toc): Promise<VFile<Toc>> {
        const vfile = new VFile<Toc>(file, copyJson(toc), dump);

        await getHooks(this).Dump.promise(vfile);

        return vfile;
    }

    private async load(file: NormalizedPath): Promise<Toc | undefined> {
        // There is no error. We really skip toc processing, if it was processed previously in any way.
        // For example toc can be processed as include of some other toc.
        if (this.relations.hasNode(file)) {
            return (this.relations.getNodeData(file) as GraphTocData).data;
        }

        this.logger.proc(file);

        const defer = new Defer<Toc | undefined>();

        this.relations.addNode(file, {type: 'toc', data: defer.promise});

        defer.promise.then((result) => {
            if (this.relations.hasNode(file)) {
                this.relations.setNodeData(file, {type: 'toc', data: result});
            }
        });

        const context: LoaderContext = this.loaderContext(file);
        const content = await read(this.run, file);

        content.path = file;

        if (this.shouldSkip(content)) {
            defer.resolve(undefined);
            return undefined;
        }

        const toc = (await loader.call(context, content)) as Toc;

        await getHooks(this).Loaded.promise(toc);

        // This looks how small optimization, but there was cases when toc is an array...
        // This is not that we expect.
        if (toc.href || toc.items?.length) {
            await this.addEntries(file, toc);
            await this.restrictAccess(file, toc);
        }

        defer.resolve(toc);

        return defer.promise;
    }

    @bounded
    private async include(path: RelativePath, include: IncludeInfo): Promise<Toc | undefined> {
        const file = normalizePath(path);

        this.relations.addNode(file);
        this.relations.setNodeData(file, {type: 'source', data: undefined});
        this.relations.addNode(include.from);
        this.relations.addDependency(include.from, file);

        this.logger.proc(file);

        const context: LoaderContext = this.loaderContext(file, include);
        const content = include.content || (await read(this.run, file, include.from));

        if (this.shouldSkip(content)) {
            return undefined;
        }

        if (isMergeMode(include)) {
            const from = normalizePath(dirname(file));
            const to = normalizePath(dirname(include.base));

            context.vars = this.vars.for(include.base);
            context.path = context.path.replace(from, to) as NormalizedPath;
            context.from = include.from;

            const files = await this.run.copy(
                normalizePath(join(this.run.input, from)) as AbsolutePath,
                normalizePath(join(this.run.input, to)) as AbsolutePath,
                [basename(file), '**/toc.yaml'],
            );

            for (const pair of files) {
                const [from, to] = pair.map((path) =>
                    normalizePath(relative(this.run.input, path)),
                );
                const sourcePath = this.meta.get(from).sourcePath || from;
                this.meta.add(to, {sourcePath});
                this.logger.copy(pair[0], pair[1]);
            }
        }

        const toc = (await loader.call(context, content)) as Toc;

        await getHooks(this).Included.promise(toc, include);

        return toc;
    }

    /**
     * Visits all passed items. Applies actor to each item.
     * Then applies actor to each item in actor result.items.
     * Returns actor results.
     * DFS
     */
    private async _walkItems<T extends WithItems<T>>(
        items: T[] | undefined,
        actor: (
            item: T,
            context: WalkStepContext,
        ) => Promise<WalkStepResult<T>> | WalkStepResult<T>,
        options: WalkOptions<T>,
        context: Hash = {},
    ): Promise<T[] | undefined> {
        const {accept} = options;

        if (!items || !items.length) {
            return items;
        }

        const results: T[] = [];

        for (const item of items) {
            const itemContext: WalkStepContext = {...context};

            const result = (accept(item) ? await actor(item, itemContext) : item) as T &
                Record<'items', unknown>;

            if (result) {
                results.push(...([] as T[]).concat(result));

                if (own(result, 'items')) {
                    // Sometime users defines items as object (one item) instead of array of one item.
                    if (!Array.isArray(result.items) && result.items) {
                        result.items = ([] as T[]).concat(result.items);
                    }

                    if (result.items?.length) {
                        result.items = await this._walkItems(
                            result.items,
                            actor,
                            options,
                            itemContext,
                        );
                    }
                }
            }
        }

        return results;
    }

    private shouldSkip(toc: RawToc) {
        const {ignoreStage} = this.config;
        if (toc.stage && ignoreStage.length && ignoreStage.includes(toc.stage)) {
            return true;
        }

        return false;
    }

    private async addEntries(path: NormalizedPath, toc: Toc) {
        await this.walkEntries([toc as unknown as EntryTocItem], (item) => {
            const entryPath = normalizePath(join(dirname(path), item.href));
            this.relations.addNode(entryPath, {type: 'entry', data: undefined});
            this.relations.addDependency(toc.path, entryPath);

            return item;
        });
    }

    private async restrictAccess(path: NormalizedPath, toc: Toc) {
        await this.walkItems(
            [toc as unknown as RawTocItem],
            (item, context: RestrictedAccessContext) => {
                if (own<string | string[]>(item, 'restricted-access')) {
                    let itemAccess = ([] as string[]).concat(item['restricted-access'] || []);

                    const contextAccess: string[][] =
                        (context['restricted-access'] as string[][]) ?? [];
                    if (contextAccess.some(isEqualAccess(itemAccess.sort().join(',')))) {
                        itemAccess = [];
                    }

                    if (itemAccess.length > 0) {
                        context['restricted-access'] = [...contextAccess, itemAccess];
                    }
                }

                if (context['restricted-access']?.length && isEntryItem(item)) {
                    const href = normalizePath(join(dirname(path), item.href));
                    this.meta.add(href, {
                        'restricted-access': context['restricted-access'],
                    });
                }

                return item;
            },
        );

        return toc;
    }

    private loaderContext(path: NormalizedPath, {from, mode, base}: Partial<IncludeInfo> = {}) {
        return {
            path,
            from: from || path,
            mode,
            base,
            vars: this.vars.for(path),
            toc: this,
            logger: this.logger,
            include: this.include,
            settings: {
                conditions: this.config.template.features.conditions,
                substitutions: this.config.template.features.substitutions,
            },
            options: {
                removeHiddenItems: this.config.removeHiddenTocItems,
                removeEmptyItems: this.config.removeEmptyTocItems,
                skipMissingVars: this.options.skipMissingVars,
                mode: this.options.mode,
            },
        };
    }
}

async function read(run: Run, path: RelativePath, from?: string): Promise<RawToc> {
    try {
        const source = normalizePath(join(run.input, path)) as AbsolutePath;
        return load((await run.read(source)) || '{}') as RawToc;
    } catch (error) {
        throw new Error(dedent`
            Unable to resolve ${path}${from ? ' from ' + from : ''}.
            Original error:
                ${errorMessage(error)}
        `);
    }
}

function isEqualAccess(match: string) {
    return function (access: string[]) {
        return access.sort().join() === match;
    };
}
