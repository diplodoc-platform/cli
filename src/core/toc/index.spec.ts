import type {RunSpy} from '~/commands/build/__tests__';
import type {RawToc, Toc} from './types';

import {join} from 'node:path';
import {describe, expect, it, vi} from 'vitest';
import {when} from 'vitest-when';
import {dump} from 'js-yaml';
import {dedent} from 'ts-dedent';

import {setupRun} from '~/commands/build/__tests__';

import {TocService, TocServiceConfig} from './TocService';
import {getHooks} from './hooks';

type Options = DeepPartial<TocServiceConfig>;

vi.mock('../vars/VarsService');

function setupService(options: Options = {}) {
    const run = setupRun({
        ignoreStage: [],
        removeHiddenTocItems: false,
        ...options,
        template: {
            enabled: true,
            ...(options.template || {}),
            features: {
                conditions: true,
                substitutions: true,
                ...((options.template || {}).features || {}),
            },
        },
    });
    const toc = new TocService(run);

    return {run, toc};
}

function mockData(run: RunSpy, content: string, vars: Vars, files: Files, copy: Copy) {
    when(run.vars.for)
        .calledWith('toc.yaml' as NormalizedPath)
        .thenReturn(vars);

    when(run.read).calledWith(join(run.input, './toc.yaml')).thenResolve(content);

    for (const [path, content] of Object.entries(files)) {
        when(run.read)
            .calledWith(join(run.input, path))
            .thenResolve(content as string);
    }

    for (const [from, to] of copy) {
        when(run.copy)
            .calledWith(join(run.input, from), join(run.input, to), expect.anything())
            .thenResolve([]);
    }
}

type Vars = Hash<string | number>;
type Files = Hash<string>;
type Copy = [RelativePath, RelativePath][];
function test(
    content: string,
    options: Options = {},
    vars: Vars = {},
    files: Files = {},
    copy: Copy = [],
) {
    return async () => {
        const {run, toc} = setupService(options);

        mockData(run, content, vars, files, copy);

        const result = (await toc.dump('toc.yaml' as NormalizedPath)) as Toc;

        expect(dump(result)).toMatchSnapshot();
    };
}

describe('toc-loader', () => {
    it(
        'should handle simple title',
        test(dedent`
            title: Title
        `),
    );

    it(
        'should handle filter title',
        test(
            dedent`
                title:
                  - text: Title A
                    when: var == "A"
                  - text: Title B
                    when: var == "B"
            `,
            {},
            {var: 'B'},
        ),
    );

    it('should interpolate title', test(`title: Title {{var}}`, {}, {var: 'C'}));

    it(
        'should interpolate conditions in title',
        test(
            dedent`
                title: Title {% if var == "C"%} IF {% endif %}
            `,
            {},
            {var: 'C'},
        ),
    );

    it(
        'should interpolate filter title',
        test(
            dedent`
                title:
                  - text: Title A
                    when: var == "A"
                  - text: Title B
                    when: var == "B"
                  - text: Title {{var}}
                    when: var == "C"
            `,
            {},
            {var: 'C'},
        ),
    );

    it(
        'should not interpolate title if substitutions is disabled',
        test(
            dedent`
                title: Title {{var}}
            `,
            {template: {features: {substitutions: false}}},
            {var: 'C'},
        ),
    );

    it(
        'should not interpolate title if conditions is disabled',
        test(
            dedent`
                title: Title {% if var == "C"%} IF {% endif %}
            `,
            {template: {features: {conditions: false}}},
            {var: 'C'},
        ),
    );

    it(
        'should not interpolate title if both templating is disabled',
        test(
            dedent`
                title: Title {% if var == "C"%} IF {% endif %}
            `,
            {template: {features: {conditions: false, substitutions: false}}},
            {var: 'C'},
        ),
    );

    it(
        'should not filter item with accepted rule',
        test(
            dedent`
                items:
                  - name: Visible Item 1
                  - name: Visible Item {{name}}
                    when: stage == 'test'
                  - name: Visible Item 3
            `,
            {},
            {stage: 'test', name: 2},
        ),
    );

    it(
        'should filter item with declined rule',
        test(
            dedent`
                items:
                  - name: Visible Item 1
                  - name: Item {{name}}
                    when: stage == 'test'
                  - name: Visible Item 2
            `,
            {},
            {stage: 'dev'},
        ),
    );

    it(
        'should filter hidden item',
        test(
            dedent`
                items:
                  - name: Visible Item 1
                  - name: Hidden Item {{name}}
                    hidden: true
                    when: stage == 'test'
                  - name: Visible Item 2
                    when: stage == 'test'
            `,
            {removeHiddenTocItems: true},
            {stage: 'test'},
        ),
    );

    it(
        'should interpolate item name',
        test(
            dedent`
                items:
                  - name: Item {{name}}
            `,
            {},
            {name: 'C'},
        ),
    );

    it(
        'should interpolate item href',
        test(
            dedent`
                items:
                  - href: "{{file}}"
            `,
            {},
            {file: './file.md'},
        ),
    );

    it(
        'should interpolate nested item',
        test(
            dedent`
                items:
                  - name: Parent
                    items:
                      - name: Item {{name}}
                        href: "{{file}}"
            `,
            {},
            {name: 'C', file: './file.md'},
        ),
    );

    it(
        'should normalize items',
        test(
            dedent`
                items:
                  - name: Item without extension
                    href: some/href
                  - name: Item with slash
                    href: some/href/
            `,
            {},
            {},
        ),
    );

    describe('includes', () => {
        it(
            'should rebase items href for includes in link mode',
            test(
                dedent`
                    items:
                      - name: Outer Item
                        include:
                          path: _includes/core/i-toc.yaml
                          mode: link
                `,
                {},
                {},
                {
                    '_includes/core/i-toc.yaml': dedent`
                        items:
                          - name: Inner Item 1
                            href: item-1.md
                            items:
                              - name: Inner Sub Item 1
                                href: sub-item-1.md
                          - name: Inner Item 2
                            href: ./item-2.md
                          - name: Inner Item 3
                            href: ./sub/item-3.md
                          - name: Inner Item 4
                            href: sub/item-4.md
                          - name: Inner Item 5
                            href: ../item-5.md
                          - name: Inner Item 6
                            href: https://example.com
                          - name: Inner Item 7
                            href: //example.com
                    `,
                },
            ),
        );

        it(
            'should filter include in preview stage',
            test(
                dedent`
                    items:
                      - name: Common item
                      - name: Filtered item
                        include:
                          path: _includes/core/i-toc.yaml
                          mode: link
                `,
                {},
                {},
                {
                    '_includes/core/i-toc.yaml': dedent`
                        stage: tech-preview
                        items:
                          - name: Inner Item 1
                    `,
                },
            ),
        );

        it(
            'should merge includes in link mode',
            test(
                dedent`
                    items:
                      - name: Outer Item
                        include:
                          path: _includes/core/i-toc.yaml
                          mode: link
                `,
                {},
                {},
                {
                    '_includes/core/i-toc.yaml': dedent`
                        items:
                          - name: Inner Item 1
                    `,
                },
            ),
        );

        it(
            'should merge includes in flat link mode',
            test(
                dedent`
                    items:
                      - include:
                          path: _includes/core/i-toc.yaml
                          mode: link
                `,
                {},
                {},
                {
                    '_includes/core/i-toc.yaml': dedent`
                        items:
                          - name: Inner Item 1
                    `,
                },
            ),
        );

        it(
            'should merge deep includes in link mode',
            test(
                dedent`
                    items:
                      - name: Outer Item
                        include:
                            path: _includes/core/i-toc.yaml
                            mode: link
                `,
                {},
                {},
                {
                    '_includes/core/i-toc.yaml': dedent`
                        items:
                          - name: Inner Item 1
                            href: item-1.md
                          - name: Inner Item 2
                            include:
                              path: ../lib/i-toc.yaml
                              mode: link
                    `,
                    '_includes/lib/i-toc.yaml': dedent`
                        items:
                          - name: Inner Lib Item 1
                            href: item-1.md
                    `,
                },
            ),
        );

        it(
            'should merge deep includes in merge mode',
            test(
                dedent`
                    items:
                      - name: Outer Item
                        include:
                            path: _includes/core/toc.yaml
                            mode: link
                `,
                {},
                {},
                {
                    '_includes/core/toc.yaml': dedent`
                        items:
                          - name: Inner Item 1
                            href: item-1.md
                          - name: Inner Item 2
                            include:
                              path: ../merge/i-toc.yaml
                              mode: merge
                    `,
                    '_includes/core/sub/toc.yaml': dedent`
                        items:
                          - name: Inner Sub Item 1
                            href: sub-item-1.md
                    `,
                    '_includes/merge/i-toc.yaml': dedent`
                        items:
                          - name: Inner Merge Item 1
                            href: merge-item-1.md
                          - include:
                              path: ../deep-merge/i-toc.yaml
                              mode: merge
                          - include:
                              path: ./sub/toc.yaml
                              mode: merge
                    `,
                    '_includes/deep-merge/i-toc.yaml': dedent`
                        items:
                          - name: Inner Deep Merge Item 1
                            href: deep-merge-item-1.md
                    `,
                },
                [
                    ['_includes/merge', '_includes/core'],
                    ['_includes/deep-merge', '_includes/core'],
                    ['_includes/core/sub', '_includes/core'],
                ] as [RelativePath, RelativePath][],
            ),
        );
    });

    describe('includers', () => {
        it('should throw on unregistered includer', async () => {
            await expect(
                test(
                    dedent`
                    items:
                      - name: Common item
                      - include:
                          path: _includes/core/i-toc.yaml
                          mode: link
                          includers:
                            - name: unknown
                `,
                    {},
                    {},
                    {
                        '_includes/core/i-toc.yaml': dedent`
                        items:
                          - name: Inner Item 1
                    `,
                    },
                ),
            ).rejects.toThrow(`Includer with name 'unknown' is not registered.`);
        });

        it('should handle registered includer', async () => {
            const {run, toc} = setupService({});
            const content = dedent`
                items:
                  - name: Common item
                  - include:
                      path: _includes/core/i-toc.yaml
                      mode: link
                      includers:
                        - name: expected
            `;
            const files = {
                '_includes/core/i-toc.yaml': dedent`
                    items:
                      - name: Inner Item 1
                `,
            };

            mockData(run, content, {}, files, []);

            getHooks(toc)
                .Includer.for('expected')
                .tap('Tests', (toc) => ({
                    ...toc,
                    stage: 'test',
                }));

            const result = (await toc.dump('toc.yaml' as NormalizedPath)) as Toc;

            expect(dump(result)).toMatchSnapshot();
        });

        it('should fix include path', async () => {
            expect.assertions(2);

            const {run, toc} = setupService({});
            const content = dedent`
                items:
                  - name: Common item
                  - include:
                      path: _includes/core
                      mode: link
                      includers:
                        - name: expected
            `;
            const files = {};

            mockData(run, content, {}, files, []);

            getHooks(toc)
                .Includer.for('expected')
                .tap('Tests', (toc, options) => {
                    expect(options).toMatchObject({
                        path: '_includes/core/toc.yaml',
                    });

                    return {
                        ...toc,
                        stage: 'test',
                    };
                });

            const result = (await toc.dump('toc.yaml' as NormalizedPath)) as Toc;

            expect(dump(result)).toMatchSnapshot();
        });

        it('should pass extra params to includer', async () => {
            expect.assertions(2);

            const {run, toc} = setupService({});
            const content = dedent`
                items:
                  - name: Common item
                  - include:
                      path: _includes/core
                      mode: link
                      includers:
                        - name: expected
                          field: value
            `;
            const files = {};

            mockData(run, content, {}, files, []);

            getHooks(toc)
                .Includer.for('expected')
                .tap('Tests', (toc, options) => {
                    expect(options).toMatchObject({
                        path: '_includes/core/toc.yaml',
                        field: 'value',
                    });

                    return {
                        ...toc,
                        stage: 'test',
                    };
                });

            const result = (await toc.dump('toc.yaml' as NormalizedPath)) as Toc;

            expect(dump(result)).toMatchSnapshot();
        });

        it('should merge includer toc to parent', async () => {
            const {run, toc} = setupService({});
            const content = dedent`
                items:
                  - name: Common item
                  - include:
                      path: _includes/core
                      mode: link
                      includers:
                        - name: expected
                          field: value
            `;
            const files = {};

            mockData(run, content, {}, files, []);

            getHooks(toc)
                .Includer.for('expected')
                .tap('Tests', (toc) => {
                    return {
                        ...toc,
                        stage: 'test',
                        items: [{name: 'Includer item 1'}],
                    } as RawToc;
                });

            const result = (await toc.dump('toc.yaml' as NormalizedPath)) as Toc;

            expect(dump(result)).toMatchSnapshot();
        });
    });
});
