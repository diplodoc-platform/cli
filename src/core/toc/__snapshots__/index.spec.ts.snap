// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toc-loader > includers > should fix include path 1`] = `
"items:
  - name: Common item
path: toc.yaml
"
`;

exports[`toc-loader > includers > should handle registered includer 1`] = `
"items:
  - name: Common item
path: toc.yaml
"
`;

exports[`toc-loader > includers > should merge includer toc to parent 1`] = `
"items:
  - name: Common item
  - name: Includer item 1
path: toc.yaml
"
`;

exports[`toc-loader > includers > should pass extra params to includer 1`] = `
"items:
  - name: Common item
path: toc.yaml
"
`;

exports[`toc-loader > includes > should merge deep includes in link mode 1`] = `
"items:
  - name: Outer Item
    items:
      - name: Inner Item 1
        href: _includes/core/item-1.md
      - name: Inner Item 2
        items:
          - name: Inner Lib Item 1
            href: _includes/lib/item-1.md
path: toc.yaml
"
`;

exports[`toc-loader > includes > should merge deep includes in merge mode 1`] = `
"items:
  - name: Outer Item
    items:
      - name: Inner Item 1
        href: _includes/core/item-1.md
      - name: Inner Item 2
        items:
          - name: Inner Merge Item 1
            href: _includes/core/merge-item-1.md
          - name: Inner Deep Merge Item 1
            href: _includes/core/deep-merge-item-1.md
          - name: Inner Sub Item 1
            href: _includes/core/sub-item-1.md
path: toc.yaml
"
`;

exports[`toc-loader > includes > should merge includes in flat link mode 1`] = `
"items:
  - name: Inner Item 1
path: toc.yaml
"
`;

exports[`toc-loader > includes > should merge includes in link mode 1`] = `
"items:
  - name: Outer Item
    items:
      - name: Inner Item 1
path: toc.yaml
"
`;

exports[`toc-loader > includes > should rebase items href for includes in link mode 1`] = `
"items:
  - name: Outer Item
    items:
      - name: Inner Item 1
        href: _includes/core/item-1.md
        items:
          - name: Inner Sub Item 1
            href: _includes/core/sub-item-1.md
      - name: Inner Item 2
        href: _includes/core/item-2.md
      - name: Inner Item 3
        href: _includes/core/sub/item-3.md
      - name: Inner Item 4
        href: _includes/core/sub/item-4.md
      - name: Inner Item 5
        href: _includes/item-5.md
      - name: Inner Item 6
        href: https://example.com
      - name: Inner Item 7
        href: //example.com
path: toc.yaml
"
`;

exports[`toc-loader > should filter hidden item 1`] = `
"items:
  - name: Visible Item 1
  - name: Visible Item 2
path: toc.yaml
"
`;

exports[`toc-loader > should filter item with declined rule 1`] = `
"items:
  - name: Visible Item 1
  - name: Visible Item 2
path: toc.yaml
"
`;

exports[`toc-loader > should handle filter title 1`] = `
"title: Title B
path: toc.yaml
"
`;

exports[`toc-loader > should handle simple title 1`] = `
"title: Title
path: toc.yaml
"
`;

exports[`toc-loader > should interpolate conditions in title 1`] = `
"title: 'Title  IF '
path: toc.yaml
"
`;

exports[`toc-loader > should interpolate filter title 1`] = `
"title: Title C
path: toc.yaml
"
`;

exports[`toc-loader > should interpolate item href 1`] = `
"items:
  - href: file.md
path: toc.yaml
"
`;

exports[`toc-loader > should interpolate item name 1`] = `
"items:
  - name: Item C
path: toc.yaml
"
`;

exports[`toc-loader > should interpolate nested item 1`] = `
"items:
  - name: Parent
    items:
      - name: Item C
        href: file.md
path: toc.yaml
"
`;

exports[`toc-loader > should interpolate title 1`] = `
"title: Title C
path: toc.yaml
"
`;

exports[`toc-loader > should load restricted-access as array 1`] = `
"name: root
restricted-access:
  - admin
  - user
items:
  - name: Item without access
    href: some/href.md
path: toc.yaml
"
`;

exports[`toc-loader > should load restricted-access as array on two level 1`] = `
"name: root
restricted-access:
  - admin
  - user
items:
  - name: Item without access
    restricted-access:
      - userA
      - userB
    href: some/href.md
path: toc.yaml
"
`;

exports[`toc-loader > should load restricted-access as string 1`] = `
"name: root
restricted-access: admin
items:
  - name: Item without access
    href: some/href.md
path: toc.yaml
"
`;

exports[`toc-loader > should normalize items 1`] = `
"items:
  - name: Item without extension
    href: some/href.md
  - name: Item with slash
    href: some/href/index.yaml
path: toc.yaml
"
`;

exports[`toc-loader > should not filter item with accepted rule 1`] = `
"items:
  - name: Visible Item 1
  - name: Visible Item 2
  - name: Visible Item 3
path: toc.yaml
"
`;

exports[`toc-loader > should not interpolate title if both templating is disabled 1`] = `
"title: Title {% if var == "C"%} IF {% endif %}
path: toc.yaml
"
`;

exports[`toc-loader > should not interpolate title if conditions is disabled 1`] = `
"title: Title {% if var == "C"%} IF {% endif %}
path: toc.yaml
"
`;

exports[`toc-loader > should not interpolate title if substitutions is disabled 1`] = `
"title: Title {{var}}
path: toc.yaml
"
`;
